### PR TITLE
kernel-install: improve uki handling

### PR DESCRIPTION
**Pull request by kraxel**

### CODE REVIEW

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml:227]: Not entirely sure that's the best solution, since this makes this harder to configure. I think a solution like mentioned in [this comment](https://github.com/systemd/systemd/issues/25822#issuecomment-1364105685) would be preferable, since having UKIs with type 1 entries might become a usecase down the road.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]: `$BOOT` is used here, because that's the language used in [BLS](https://uapi-group.org/specifications/specs/boot_loader_specification/#the-boot-partition-placeholder)

File: src/kernel-install/90-uki-copy.install
Comment (behrmann) [inline src/kernel-install/90-uki-copy.install]: `kernel-install` sets `KERNEL_INSTALL_BOOT_ROOT` (see `BOOT_ROOT_SEARCH` there). Also, it is configurable, so there might be cases where `KERNEL_INSTALL_BOOT_ROOT`, when set from the outside, and `bootctl --print-esp-path` might not agree.

File: src/kernel-install/90-uki-copy.install
Comment (behrmann) [inline src/kernel-install/90-uki-copy.install]: Even with autodetection of UKIs, does it make sense to remove the convention?

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml:227]: If we want make this configurable I don't think this should be `layout=...` but a new, separate switch.

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: `$BOOT` and `$ESP` are not the same thing.

File: src/kernel-install/90-uki-copy.install
Comment (kraxel) [reply src/kernel-install/90-uki-copy.install]: Yes.  That is the whole point.  `$KERNEL_INSTALL_BOOT_ROOT` might point to something which is not the ESP.  On a typical fedora install for example you have `BOOT=/boot` and `ESP=/boot/efi`.  The unified kernel needs to go to the ESP though.

File: src/kernel-install/90-uki-copy.install
Comment (kraxel) [reply src/kernel-install/90-uki-copy.install]: It looked like a sanity check to me.  With autodetection we know at this point that the given kernel image actually is a UKI (which implies it is a EFI PE executable), so the check IMHO doesn't make much sense any more.

We could continue enforce the `*.efi` naming convention.  Not sure what the point is, but if you prefer it that way I can drop the change.  I don't care much either way.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml:227]: I'm not sure it makes sense to add another option, when it leads back to `KERNEL_INSTALL_LAYOUT` having a single valid option, which historically meant BLS type 1.

File: src/kernel-install/90-uki-copy.install
Comment (behrmann) [reply src/kernel-install/90-uki-copy.install]: But the ESP can be mounted at `/boot` and [BLS](https://uapi-group.org/specifications/specs/boot_loader_specification/#the-boot-partition-placeholder) actually encourages this. 
> In the text below, the placeholder `$BOOT` will be used to refer to the partition determined as follows:
> 3. Otherwise, on disks with a GPT partition table: → the ESP

In the logic in `kernel-install` `BOOT_ROOT` will be either `/boot`, `/efi` or `/boot/efi` (although the latter is something that is not recommended), i.e. this will already be set to the correct value outside of the plugin.

My point above is still different, though: `KERNEL_INSTALL_BOOT_ROOT` can be set externally for the invocation of `kernel-install` via the `BOOT_ROOT` environment variable. It will break backward compat if that is ignored and something else is used in its place.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml]: They can be though. If I remember the spec correctly, UKis also don't need to be in the ESP, they can also be in `XBOOTLDR`.

**Comment by poettering** [comment]
would really love @keszybz' input on this

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml:227]: UKI and non-UKI kernels are different things, having a single option for both doesn't make sense to me.  Specifically the problem with `layout=uki` is that installing a non-UKI kernel in `/EFI/Linux` will simply not work in most cases because the initrd is missing then ...

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: Yes, `$BOOT` and `$ESP` can be identical.  That is the easy case.  But what if they are not?  One advantage of UKIs is that they can also be booted without any bootloader, directly from the firmware, and you loose that if they are placed on a filesystem the firmware can't read.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml:227]: That is true, but if you choose to do UKIs and you are installing something else there, than there's already an issue with whatever tool generates your UKIs, no? Maybe I'm misunderstanding something.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml]: That is completely true and therefore the [spec says](https://uapi-group.org/specifications/specs/boot_loader_specification/#the-partitions):
> For systems where the firmware is able to read file systems directly, the ESP must — and the MBR boot and GPT XBOOTLDR partition should — be a file system readable by the firmware. For most systems this means VFAT (16 or 32 bit). Applications accessing both partitions should hence not assume that fancier file system features such as symlinks, hardlinks, access control or case sensitivity are supported.

This is also the reason why the ESP should not be mounted to `/boot/efi`, so that `/boot` and the ESP are not nested mounts and can be automounted.

Also, UKIs can be in `XBOOTLDR`, so writing `ESP` is too narrow.

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml:227]: That may work on your local machine.  Doing the switch to UKIs at distro-level is not going to work that way.  We will see UKIs and non-UKIs living side-by-side for quite a while, maybe forever.  Fedora will ship kernel rpm packages for both UKI and non-UKI kernels.  So having UKI and non-UKI kernels installed at the same time can easily happen and there are valid reasons to do that.  For testing and development.  Or just temporary when switching a machine from one kind of kernels to the other (install new, reboot, uninstall old).

More details: https://fedoraproject.org/wiki/Changes/Unified_Kernel_Support_Phase_1

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: Reality often ignores the BLS recommendations.  Anaconda (the Fedora Installer) even throws errors in case I ask it to use vfat for /boot.

I can see that installing to `$BOOT` by default makes sense for consistency reasons, but I guess I want a config switch then so I have the option to force using `$ESP` instead.

Comment (behrmann) [reply man/kernel-install.xml]: You already do, you can set `BOOT_ROOT` for the invocation of `kernel-install`.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml:227]: Yes, having UKIs and non-UKIs installed at the same time is valid, at least for switching from one to the other, but I'm not sure how common regular intermixing of UKI and non-UKI kernel packages will be in the wild (I'd assume you'd settle for one in the end). For dev purposes you could already point `kernel-install` to a different config altogether (via `KERNEL_INSTALL_CONF_ROOT`).

What I'm trying to get at is that so far `layout=` was the one dimension that parametrised the behaviour of `kernel-install` and for the longest time there was only really a single valid state (`bls`, besides the invalid or agnostic state of "I don't care what the layout is"). This state implements BLS type 1. The idea for the `uki` layout was to have an explicit switch to change that behaviour by introducing another valid state. You propose to reduce valid states back to a single one and introduce another dimension to parametrise the behaviour. At least that is my understanding, please correct me if I'm wrong. I'm not convinced by this, since if there every are to be multiple layouts, this makes for more combinations that need to be implemented an tested. I think if the idea is to encode multiple different behaviours, it is better to do this via different layouts. 

I think, a better solution would be if `kernel-install` could do multiple layouts for a single run of `kernel-install` (this could be hacked now, by running kernel-install with different configs, i.e. different values for layout or by specifying the plugins to run via `KERNEL_INSTALL_PLUGINS`).

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: Same problem we have in the other discussion thread wrt. `layout=...`.  I can stick that into `install.conf`, but it'll be applied to both UKI and non-UKI kernels, which is not what I want.

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml:227]: There is `layout=bls` which will effectively activate `90-loaderentry`, which installs the kernel according to BLS to `$BOOT/$MACHINE_ID/$KERNELVER/linux` and creates a BLS type1 config snippet for it.

There is `layout=other` (or layout being unset) where the systemd-shipped scripts don't do anything.  In case of Fedora a script dropped by the grub rpms runs in that case and installs the kernel and updates the config.  It's actually a BLS type1 entry too (fedora grub is patched to understand that), but kernel + initrd are placed elsewhere (`$BOOT/vmlinuz-$KERNELVER`).  This is Fedoras default behavior today.

So there are already two possible options for `layout=`.

For UKIs the default should be to install them to `/EFI/Linux`, i.e. BLS type 2.  But flipping the default for non-UKI kernels to BLS type 2 doesn't make sense at all.  Installing UKIs as type 1 might make sense though if you want add command line options, so we should probably allow that.  So we could have:

```
layout={bls,other}
uki_layout={uki,bls,other}
```

Oh, and expecting users and developers work with different config files and/or setting `KERNEL_INSTALL_*` is not going to fly in practice.  People will get it wrong all the time then.

File: src/kernel-install/90-uki-copy.install
Comment (keszybz) [reply src/kernel-install/90-uki-copy.install]: https://uapi-group.org/specifications/specs/boot_loader_specification/#type-2-efi-unified-kernel-images is explicit: "the files must have the extension `.efi`". So let's keep the check for now.

File: man/kernel-install.xml
Comment (keszybz) [reply man/kernel-install.xml:227]: We *will* have both types of images installed on the same system, so the tools should support flexible operation. The two scenarios that are "obvious":
1. traditional BLS type#1 installations with an entry file and linux and initrd
2. pre-built UKI delivered via a package

but there are some plans to also support:

3. a mode where you reference a UKI and create a BLS type#1 entry for it with some override of options
4. building a UKI on demand

Case 3. is useful when you want to override options and don't want to pay the price (or maybe even can't) of having two sets of kernel+initrd installed. Case 4. is useful when you want to use boot measurement or attestation but the distro doesn't give you the right kernel or you are building something locally.

I'd like to work backwards from those scenarios and figure out what the interface should be. But frankly, I don't quite see how each case would work yet.

File: man/kernel-install.xml
Comment (keszybz) [reply man/kernel-install.xml]: Maintainers of Anaconda refusing to drop the misguided and pointless check against vfat is hard to explain. There are no technical reasons for this. So we really shouldn't use Anaconda as justification for anything.

There are good reasons to allow installing to XBOOTLDR: the ESP might be too small. It won't work in all cases, but in some it certainly will, so it's a useful option to have.

File: src/kernel-install/90-uki-copy.install
Comment (keszybz) [reply src/kernel-install/90-uki-copy.install]: Yeah, this change just makes this worse. Please just let `$BOOT_ROOT` do its job.

File: src/kernel-install/90-uki-copy.install
Comment (keszybz) [inline src/kernel-install/90-uki-copy.install:53]: Yeah, that's reasonable. It makes the installation procedure more robust.

**Comment by keszybz** [comment]
Instead of the changes of logic and naming, I think this series should be made much simpler: just set KERNEL_INSTALL_LAYOUT=uki if not set and the kernel is a UKI. This will make installation work out of the box for the common case with UKIs. Other uses of UKIs will still be possible, with additional config. But they are more complicated and would require additional config anyway.

File: man/kernel-install.xml
Comment (poettering) [inline man/kernel-install.xml:227]: So I think we should explicitly support a mode where type1 entries are generated for UKIs for people who want to set kernel cmdline for example. In fact in sd-boot we should explicitly support that if a type1 and a type2 entry we find share the same name we'll ignore the type2 entry, so that people can "wrap" a type2 kernel as type1 if they like, to paramaterize it further.

(of course, this will come with implications on secureboot, i.e. what will actually be allowed, but that's another question).

hence, I think three scenarios matter in the long run

1. classic non-UKI kernel package, with local initrd, installed as type1
1. UKI packaged, and installed 1:1 as type2
2. UKI packaged, and installed with a type1 maybe because /etc/kernel/cmdline was set or some other reason (maybe a devicetree file or so)
3. classic non-UKI kernel packaged, from which a local UKI is generated/signed, and installed as type1

hence maybe:

KERNEL_INSTALL_LAYOUT=bls|uki|auto|other

bls: always install as type 1. i.e. for packaged classic non-UKI kernels, install as regular type1. and for UKI kernels generate a type2 wrapper.
uki: always install as type 2. i.e. for packaged classic non-UKI kernels, generate UKI and install as UKI. for UKI kernels install 1:1
auto: try to be smart, i.e. install classic kernels as type1, and UKIs as type2. Except: if a local signing key is found in /etc/kernel/uki-sign.pem (or so), then take that as hint to generate a UKI from classic kenrels, and sign them (i.e. if people drop-in stuff upgrade classic kernels to UKI mode). And the opposite direction too: if /etc/kernel/cmdline exists, install type1 wrapper for uki with it included.

but I'd see that as the long term perspective. we should keep this in mind even if we don't implement all this right-away.

File: man/kernel-install.xml
Comment (poettering) [reply man/kernel-install.xml]: the whole concept of `$BOOT` exists only to be used as a *the* place to drop new kernels into, hence we should use it here.

not sure anaconda matters here, given it doesn't support XBOOTLDR anyway, does it? its boot partition does not use the xbootldr partition type, does it?

Comment (kraxel) [reply man/kernel-install.xml]: FYI: @berrange made some progress on the anaconda front.

Both parted python bindings and blivet are green upstream:
https://bugzilla.redhat.com/show_bug.cgi?id=1075288

Anaconda itself is not (yet) ready:
https://bugzilla.redhat.com/show_bug.cgi?id=2160074

For now I'm doing `sfdisk` calls in `%post` in my test kickstart files.

I'm fine with using `$BOOT` by default but I need an override for use cases where I really need the UKI being placed on `$ESP`.

**Comment by kraxel** [comment]
> Instead of the changes of logic and naming, I think this series should be made much simpler: just set KERNEL_INSTALL_LAYOUT=uki if not set and the kernel is a UKI. This will make installation work out of the box for the common case with UKIs. Other uses of UKIs will still be possible, with additional config. But they are more complicated and would require additional config anyway.

Update pushed with that.  Keeping the new KERNEL_INSTALL_IMAGE_TYPE variable though, I think it will be useful for plugins (i.e. dracut plugin skipping inited generation for UKI kernels).

File: src/kernel-install/kernel-install.in
Comment (behrmann) [inline src/kernel-install/kernel-install.in]: Needs an entry in the man page.

File: src/kernel-install/kernel-install.in
Comment (behrmann) [inline src/kernel-install/kernel-install.in]: Switching the layout if none is set might break backward-compat. @poettering mentioned an `auto` layout above. Maybe something like 
```suggestion
    if [ "$KERNEL_INSTALL_IMAGE_TYPE" = "uki" ] && [ "$layout" = "auto" ]; then
```
and switching the default to `auto`, would make sense, since that should allow people to still not set anything and keep the current behaviour.

File: src/kernel-install/kernel-install.in
Comment (kraxel) [reply src/kernel-install/kernel-install.in]: Sure, I'll add then when we have agreement on the code changes.

File: src/kernel-install/kernel-install.in
Comment (kraxel) [reply src/kernel-install/kernel-install.in]: Ok, and then eventually pass "auto" to the plugins?  Or should `kernel-install` decide what mode to pick then?  I think the latter would be better.

Comment (behrmann) [reply src/kernel-install/kernel-install.in]: I agree. Otherwise every plugin would need to reimplement that logic and that would only lead to problems down the line.

*Edit*: More clearly, yes, `kernel-install` should decide the mode for `auto`.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]:
```suggestion
      <para><varname>$KERNEL_INSTALL_LAYOUT=bls|uki|auto|other|...</varname> is set for the plugins to specify the
```
`auto` should at least be before `other`, because `auto` actually exists, whereas `other` is just a placeholder for whatever else people build themselves.

And since the series makes it the default it should be
```suggestion
      <para><varname>$KERNEL_INSTALL_LAYOUT=auto|bls|uki|other|...</varname> is set for the plugins to specify the
```

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]:
```suggestion
            <para>Pick layout automatically.  This will set layout to <option>uki</option> in case the kernel image passed in
```
Looking through the manpage it actually states in the manpage (in the environment variable section) that `bls` is the default for `KERNEL_INSTALL_LAYOUT`.

File: src/kernel-install/kernel-install.in
Comment (behrmann) [inline src/kernel-install/kernel-install.in]: Sorry, looking at this yesterday, I forgot about the big block at around line 250 that checks for `[ -z "$layout" ]`. The man page saying the default is `bls` reminded me of that. This should remain empty and the auto handling be moved there, because it works a few other cases as well.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]:
```suggestion
            <para>Pick the layout automatically.  This inspects the kernel image and <filename>$BOOT</filename>
```

Probably minor nit, the rest so the man pages doesn't do double spaces after periods, so this shouldn't either.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]: This should be changed as well.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]:
```suggestion
            and will choose a layout based on what it finds. Leaving <varname>layout</varname> blank has the same effect. This
```
This should probably also say, that `uki` is chosen if the passed in stuff is a UKI, `bls` if `$BOOT/loader/entries.srel` or an entry token is found and `other` otherwise.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]: This is not a typo. But this misunderstanding happens a lot in the docs. iff is mathematician slang for "if and only if". Since we kept this around, this should stay, too.

File: man/kernel-install.xml
Comment (behrmann) [reply man/kernel-install.xml]: Ah, Github highlights this weirdly... I mean the "Defaults to `bls`" in 228.

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: The logic is actually more complex, and I have my doubts that it makes much sense to replicate that in the documentation.

Comment (behrmann) [reply man/kernel-install.xml]: Only slightly so, no? The content of the srel file is checked to be `bls`. I'm on the side that one should be able to know what's about to happen without having to read the code, but I agree that it can be problematic since documentation and implementation already diverged.

File: man/kernel-install.xml
Comment (kraxel) [reply man/kernel-install.xml]: Using math slang in the docs doesn't look like a good idea to me.

Care to explain what exactly the semantic difference is?  I read both versions as only kernel images named *.efi being accepted.

Comment (behrmann) [reply man/kernel-install.xml]: Just saw that this actually is the last userfacing "iff", the rest is only in comments, so this should go as well in the interest of being jargon free. A valid subsitution could be 
```suggestion
            not exist the <replaceable>KERNEL-IMAGE</replaceable> argument, only if it has a
```

Logically this means if is a one way implication and iff is bidrectional. In both cases "The file has a .efi suffix, so it will be copied to the directory" is logically valid, but "The file has been copied to the directory, so it must have a .efi suffix" is only true in the latter not the former.

File: man/kernel-install.xml
Comment (keszybz) [reply man/kernel-install.xml:227]: Heh, this is pretty much what I wrote above: https://github.com/systemd/systemd/pull/26119#discussion_r1083972379. We even used the same numbering scheme.

> `KERNEL_INSTALL_LAYOUT=bls|uki|auto|other`

Yeah.

> Except: if a local signing key is found in /etc/kernel/uki-sign.pem (or so), then take that as hint to generate a UKI from classic kenrels, and sign them (i.e. if people drop-in stuff upgrade classic kernels to UKI mode).

Mmmm, maybe. Sounds complicated. I think we should aim for the situation where distros sign kernels, and special cases like local signing can be handled with explicit configuration.

File: src/boot/bootctl-uki.c
Comment (keszybz) [inline src/boot/bootctl-uki.c]: Shouldn't those two just say `"File is smaller than DOS magic | DOS header (%zu bytes)", items`? "Read error" is misleading for the user, there is no "read error" here.

File: src/kernel-install/kernel-install.in
Comment (keszybz) [inline src/kernel-install/kernel-install.in]: Capital "K" please.

**Review by keszybz — APPROVED** [review]
LGTM.

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]: Missing an article
```suggestion
            <para>Pick the layout automatically. If the kernel is a UKI set layout to <option>uki</option>. If
```
or
```suggestion
            <para>Pick a layout automatically. If the kernel is a UKI set layout to <option>uki</option>. If
```

File: man/kernel-install.xml
Comment (behrmann) [inline man/kernel-install.xml]:
```suggestion
            not default to <option>bls</option> if <filename>$BOOT/loader/entries.srel</filename> with content <literal>type1</literal> or
            <filename>$BOOT/<replaceable>ENTRY-TOKEN</replaceable></filename> exists, and
```
since the `entries.srel` file is specifically mentioned in the spec.

**Review by behrmann — COMMENTED** [review]
Otherwise LGTM, just two last nits

**Review by behrmann — APPROVED** [review]

**Comment by keszybz** [comment]
In the meantime #26236 fixed the installation issue in a slightly different way. I pushed a rebase with that patch dropped.

**Comment by bluca** [comment]
is this related?

```
 360/1399 systemd / test-kernel-install                                                           FAIL              0.07s   exit status 1
>>> PROJECT_BUILD_ROOT=/home/runner/work/systemd/systemd/build MALLOC_PERTURB_=185 PATH=/home/runner/work/systemd/systemd/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin SYSTEMD_LANGUAGE_FALLBACK_MAP=/home/runner/work/systemd/systemd/src/locale/language-fallback-map /home/runner/work/systemd/systemd/src/kernel-install/test-kernel-install.sh /home/runner/work/systemd/systemd/build/kernel-install src/kernel-install/90-loaderentry.install
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stdout:
Reading /tmp/test-kernel-install.b0G9tSvJBB/sources/install.conf…
/tmp/test-kernel-install.b0G9tSvJBB/sources/install.conf configures layout=bls
/tmp/test-kernel-install.b0G9tSvJBB/sources/install.conf configures initrd_generator=none
MACHINE_ID=3e0484f3634a418b8e6a39e8828b03e3 set via environment
BOOT_ROOT=/tmp/test-kernel-install.b0G9tSvJBB/boot set via environment
entry-token "the-token" acquired from /tmp/test-kernel-install.b0G9tSvJBB/sources/entry-token
Entry-token candidates: the-token
/tmp/test-kernel-install.b0G9tSvJBB/boot/the-token not found…
/tmp/test-kernel-install.b0G9tSvJBB/boot/loader/entries not found…
stderr:
Unknown command verb kernel-identify.

```

**Comment by kraxel** [comment]
> Looks like [this line](https://github.com/systemd/systemd/pull/26119/files#diff-4458868cf2352dde45db167301235ad7e22f3508ae16f5389e5cf12899fecbb2R255) is not picking up the correct `bootctl` from the build? The commits introduced `kernel-identify` are on this branch at least.

The normal `ninja test` run uses the correct (fresh built) bootctl binary.  Which is needed because of the bootctl bugfix in this series.  The commits introducing `kernel-identify` are actually in v253, so the failing test must use 252 or older bootctl version for some reason ...

Pushed rebased branch (without changes) to re-trigger CI.

File: src/boot/bootctl-uki.c
Comment (poettering) [inline src/boot/bootctl-uki.c]: to make this less surprising, maybe change first argument to `&dos.Magic`, too.

might confuse static checkers otherwise.

Comment (poettering) [reply src/boot/bootctl-uki.c]: or alternatively:

```c
items = fread(&dos, 1, endoffsetof_field(DosFileHeader, Magic), uki);
```

File: src/boot/bootctl-uki.c
Comment (poettering) [inline src/boot/bootctl-uki.c]: and then instead of rewinding here maybe do:

```c
items = fread((uint8_t*) &dos + items, 1, sizeof(DosFileHeader) - items, uki);
```

File: src/kernel-install/90-uki-copy.install
Comment (poettering) [inline src/kernel-install/90-uki-copy.install:52]: should still go to `>&2` no?

**Comment by poettering** [comment]
lgtm

File: src/boot/bootctl-uki.c
Comment (poettering) [reply src/boot/bootctl-uki.c]: and of course the change the items == check below also to use endoffsetof_field()

File: src/kernel-install/90-uki-copy.install
Comment (kraxel) [reply src/kernel-install/90-uki-copy.install:52]: Using stdout is consistent with other places doing verbose logging.

File: src/boot/bootctl-uki.c
Comment (kraxel) [reply src/boot/bootctl-uki.c]: The pointer arithmetic here is not exactly helpful to make the code more readable.
It avoids one syscall, but this certainly isn't a performance sensitive place ...

Comment (kraxel) [reply src/boot/bootctl-uki.c]: Solved that slightly different now, do a single read but do two size checks, first magic only, then full header.

**Pull request merged**
